server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java

```diff
@@ -1011,14 +1012,25 @@ public CompletableFuture<Revision> updateTokenLevel(Author author, String appId,
                               "Update the token level: " + appId + " to " + (toBeAdmin ? "admin" : "user"),
                               () -> tokenRepo.fetch(INTERNAL_PROJECT_DOGMA, Project.REPO_DOGMA, TOKEN_JSON)
                                              .thenApply(tokens -> {
-                                                 final JsonPointer adminPath = JsonPointer.compile(
-                                                         "/appIds" + encodeSegment(appId) +
-                                                         "/admin");
+                                                 final Tokens tokens0 = tokens.object();
+                                                 final Token token = tokens0.get(appId);
+                                                 if (token == null) {
+                                                     throw new TokenNotFoundException("App ID: " + appId);
+                                                 }
+                                                 if (toBeAdmin == token.isAdmin()) {
+                                                     throw new IllegalArgumentException(
+                                                             "The token is already " +
+                                                             (toBeAdmin ? "admin" : "user"));
+                                                 }
```

Comment:
Question) The serialized format of `tokens.json` seems the same. Did you change to validate this? I am asking because this could override the previous state for multiple updates. For example, one deprecates a token, and another makes this token admin. This may not happen in the real world as the permission is only granted to admins.

Reply:
> The serialized format of tokens.json seems the same.

That is true except the token now deserializes from both: `{"appId": "..", "admin": true, ...}` and `{"appId": "..", "systemAdmin": true, ...}`. This is the preliminary work for #1054.
Let's imagine that we rolling-update with #1054. When a new structured token (with `systemAdmin`) is committed, the old servers will fail to read the token because it cannot recognize `systemAdmin` property.

Reply:
I understood that the JSON pointer to `/appIds/{appId}/admin` should be `/appIds/{appId}/systemAdmin` when #1054 is merged.

Reply:
After this PR gets merged, the JSON pointer does not point `/appIds/{appId}/admin` anymore. It replaces the whole `Token` at `/appIds/{appId}`. 😉